File: two_pointers/merge_sorted_array.py
from typing import List

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        nums1Index = m - 1
        nums2Index = n - 1
        index = len(nums1) - 1

        while nums2Index >= 0 and nums1Index >= 0:
            if nums1[nums1Index] > nums2[nums2Index]:
                nums1[index] = nums1[nums1Index]
                nums1Index -= 1
            else:
                nums1[index] = nums2[nums2Index]
                nums2Index -= 1

            index -= 1

        if nums2Index > -1:
            while nums2Index >= 0:
                nums1[index] = nums2[nums2Index]
                nums2Index -= 1
                index -= 1
        
        return nums1

nums1 = [2, 0]
m = 1
nums2 = [1]
n = 1

c = Solution()

print(c.merge(nums1, 1, nums2, 1))

File: two_pointers/merge_sorted_array.js
// Q:88 - https://leetcode.com/problems/merge-sorted-array
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
var merge = function(nums1, m, nums2, n) {
    let i = m - 1, j = n - 1;
    let endIndex = m + n - 1;

    while (i >= 0 && j >= 0){
        if (nums1[i] > nums2[j]){
            nums1[endIndex] = nums1[i];
            i -= 1;
        } else {
            nums1[endIndex] = nums2[j];
            j -= 1;
        }
        
        endIndex -= 1;
    }

    if (j >  -1){
        while (j >= 0){
            nums1[endIndex] = nums2[j];
            j -= 1;
            endIndex -= 1;
        }
    }

    return nums1;
};

console.log(merge([1,2,3,0,0,0], 3, [2,5,6], 3))
console.log(merge([1], 1, [], 0))
console.log(merge([0], 0, [1], 1))

console.log(merge([4,5,6,0,0,0], 3, [1,2,3], 3))


// 1, 2, 2, 3